src/blas/impl/KokkosBlas1_iamax_impl.hpp

```diff
@@ -0,0 +1,238 @@
+	mag_type val = IPT::norm (m_x(i));
+    if( val > lmaxloc.val ) { lmaxloc.val = val; lmaxloc.loc = i; }
+  }
+};
+
+/// \brief Column-wise Iamax functor for multivectors.
+///
+/// \tparam RV 1-D output View
+/// \tparam XMV 2-D input View
+/// \tparam MagType Magnitude type
+/// \tparam SizeType Index type.  Use int (32 bits) if possible.
+template<class RV, class XMV, class MagType, class SizeType = typename XMV::size_type>
+struct MV_Iamax_FunctorVector
+{
+  typedef typename XMV::execution_space                         execution_space;
+  typedef typename XMV::memory_space                            memory_space;
+  typedef SizeType                                              size_type;
+  typedef MagType                                               mag_type;
+  typedef typename XMV::non_const_value_type                    xvalue_type;
+  typedef Kokkos::Details::InnerProductSpaceTraits<xvalue_type> IPT;
+  typedef Kokkos::MaxLoc<mag_type,size_type>                         maxloc_reducer;
+  typedef typename Kokkos::MaxLoc<mag_type,size_type>::value_type    maxloc_type;
+  typedef typename Kokkos::TeamPolicy<execution_space>::member_type  member_type;
+
+  typename XMV::const_type m_x;
+  typename RV::non_const_type m_r;
+  
+  MV_Iamax_FunctorVector (const RV& r, const XMV& x) :
+    m_r (r), m_x (x)
+  {
+    static_assert (Kokkos::Impl::is_view<RV>::value,
+                   "KokkosBlas::Impl::MV_Iamax_FunctorVector: "
+                   "R is not a Kokkos::View.");
+    static_assert (Kokkos::Impl::is_view<XMV>::value,
+                   "KokkosBlas::Impl::MV_Iamax_FunctorVector: "
+                   "X is not a Kokkos::View.");
+    static_assert (Kokkos::Impl::is_same<typename RV::value_type,
+                   typename RV::non_const_value_type>::value,
+                   "KokkosBlas::Impl::MV_Iamax_FunctorVector: "
+                   "R is const.  It must be nonconst, because it is an output "
+                   "argument (we must be able to write to its entries).");
+    static_assert (RV::rank == 1 && XMV::rank == 2,
+                   "KokkosBlas::Impl::MV_Iamax_FunctorVector: "
+                   "RV must have rank 1 and XMV must have rank 2.");
+    static_assert (Kokkos::Impl::MemorySpaceAccess< typename XMV::device_type::memory_space, typename RV::device_type::memory_space >::accessible,
+                   "KokkosBlas::Impl::MV_Iamax_FunctorVector: "
+                   "RV and XMV must have the same memory space if RV is 1-D view.");
+  }
+
+    KOKKOS_INLINE_FUNCTION
+    void operator() (const member_type &teamMember) const
+    {
+      const int lid = teamMember.league_rank();// teamId
+      const int tid = teamMember.team_rank(); // threadId
+
+      maxloc_type col_maxloc;
+      Kokkos::parallel_reduce( Kokkos::TeamThreadRange(teamMember, m_x.extent(0)), [&] (const int i, maxloc_type& thread_lmaxloc) {
+        mag_type val = IPT::norm (m_x(i,lid));
```

Comment:
@vqd8a What you are doing is that parallel reduce along a vector, which is typically large compared to the number of columns (extent(1)). Logically this can do the job but will be extremely slow comparable to a serial version. You should put the most outer parallel loop for the biggest work chunk.

Reply:
@kyungjoo-kim Thanks. I have fixed the implementation as per your suggestion. Please take a look.